src/Microsoft.VisualStudio.Threading/ThreadingTools.cs

```diff
@@ -200,21 +200,54 @@ internal static void AttachCancellation<T>(this TaskCompletionSource<T> taskComp
+                            {
+                                t.CancellationCallback?.OnCanceled();
+                            }
                         },
                         tuple,
                         useSynchronizationContext: false);
+
+                    // In certain race conditions, our continuation could execute inline. We could force it to always run
+                    // asynchronously, but then in the common case it becomes less efficient.
+                    // Instead, we will optimize for the common (no-race) case and detect if we were inlined, and if so, defer the work
+                    // to avoid making our caller block for arbitrary code since CTR.Dispose blocks for in-progress cancellation notification to complete.
                     taskCompletionSource.Task.ContinueWith(
                         (_, s) =>
                         {
                             var t = (CancelableTaskCompletionSource<T>)s;
-                            t.CancellationTokenRegistration.Dispose();
+                            if (t.ContinuationScheduled || !t.OnOwnerThread)
+                            {
+                                // We're not executing inline... Go ahead and do the work.
+                                t.CancellationTokenRegistration.Dispose();
+                            }
+                            else if (!t.CancellationToken.IsCancellationRequested) // If the CT is canceled, the CTR is implicitly disposed.
+                            {
+                                // We hit the race where the task is already completed another way,
+                                // and our continuation is executing inline with our caller.
+                                // Dispose our CTR from the threadpool to avoid blocking on 3rd party code.
+                                ThreadPool.QueueUserWorkItem(
+                                    s2 =>
+                                    {
+                                        try
+                                        {
+                                            var t2 = (CancelableTaskCompletionSource<T>)s2;
+                                            t2.CancellationTokenRegistration.Dispose();
+                                        }
+                                        catch (Exception ex)
+                                        {
+                                            // Swallow any exception.
+                                            Report.Fail(ex.Message);
+                                        }
+                                    },
+                                    s);
+                            }
                         },
                         tuple,
                         CancellationToken.None,
                         TaskContinuationOptions.ExecuteSynchronously,
                         TaskScheduler.Default);
+                    tuple.ContinuationScheduled = true;
```

Comment:
Will it matter that this is set even in the case that it's executed inline?

Reply:
No. No one reads it after that in that case. Really the only point of this property is to ascertain whether the ContinueWith method has returned, so that if the continuation executes on the same thread, we can tell whether it's an inline execution or not.